File: src/main/java/mesi/orm/persistence/SQLiteResultSetParser.java
package mesi.orm.persistence;

import mesi.orm.exception.ORMesiException;
import mesi.orm.exception.ORMesiFetchException;

import java.sql.ResultSet;
import java.sql.SQLException;

class SQLiteResultSetParser implements ResultSetParser{
    @Override
    public Object parseObjectFromResultSet(String name, ResultSet rs, Class cls) throws SQLException {

        if(cls.equals(int.class) || cls.equals(Integer.class)) {
            return rs.getInt(name);
        }
        else if(cls.equals(long.class) || cls.equals(Long.class)) {
            return rs.getLong(name);
        }
        else if(cls.equals(float.class) || cls.equals(Float.class)) {
            return rs.getFloat(name);
        }
        else if(cls.equals(double.class) || cls.equals(Double.class)) {
            return rs.getDouble(name);
        }
        else if(cls.equals(String.class)) {
            return rs.getString(name);
        }
        else {
            if(PersistenceManager.isObjectPersistent(cls)) {
                return null;
            }
            else {
                throw new ORMesiFetchException("unsupported class type");
            }
        }
    }
}

File: orm-lib/src/main/kotlin/mesi/orm/persistence/fetch/SQLiteResultSetParser.kt
package mesi.orm.persistence.fetch

import mesi.orm.exception.ORMesiException
import mesi.orm.persistence.transform.PersistentProperty
import mesi.orm.persistence.transform.PersistentPropertyType
import java.sql.ResultSet
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import kotlin.reflect.full.functions

internal class SQLiteResultSetParser : ResultSetParser{
    override fun parsePropertyFrom(persistentProperty: PersistentProperty, rs: ResultSet) : Any? {
        return when(persistentProperty.type) {
            PersistentPropertyType.BOOL -> rs.getBoolean(persistentProperty.name)
            PersistentPropertyType.LONG -> {
                when(persistentProperty.kotlinClass) {
                    Int::class -> rs.getInt(persistentProperty.name)
                    Long::class -> rs.getLong(persistentProperty.name)
                    else -> throw ORMesiException("\nInvalid integer value stored in database for ${persistentProperty.kotlinClass.simpleName}:${persistentProperty.name}")
                }
            }
            PersistentPropertyType.DOUBLE -> {
                when(persistentProperty.kotlinClass) {
                    Float::class -> rs.getFloat(persistentProperty.name)
                    Double::class -> rs.getDouble(persistentProperty.name)
                    else -> throw ORMesiException("\nInvalid float value stored in database for ${persistentProperty.kotlinClass.simpleName}:${persistentProperty.name}")
                }
            }
            PersistentPropertyType.STRING -> {
                val stringRepresentation = rs.getString(persistentProperty.name)
                if(persistentProperty.isEnum) {
                    return persistentProperty.kotlinClass.functions.find { it.name == "valueOf" }?.call(stringRepresentation)
                }
                else {
                    return stringRepresentation
                }
            }
            PersistentPropertyType.TIME -> {
                val stringRepresentation = rs.getString(persistentProperty.name)
                return LocalTime.parse(stringRepresentation, DateTimeFormatter.ISO_LOCAL_TIME)
            }
            PersistentPropertyType.DATE -> {
                val stringRepresentation = rs.getString(persistentProperty.name)
                return LocalDate.parse(stringRepresentation, DateTimeFormatter.ISO_LOCAL_DATE)
            }
            PersistentPropertyType.DATETIME -> {
                val stringRepresentation = rs.getString(persistentProperty.name)
                return LocalDateTime.parse(stringRepresentation, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            }
        }
    }
}